pgmax/bp/infer.py

```diff
@@ -97,7 +97,112 @@ def pass_fac_to_var_messages(
                 )
             )
         )
+    return ftov_msgs
+
+
+@jax.jit
+def pass_OR_fac_to_var_messages(
+    vtof_msgs: jnp.ndarray,
+    parents_states: jnp.ndarray,
+    children_states: jnp.ndarray,
+) -> jnp.ndarray:
+
+    """Passes messages from OR Factors to Variables.
+
+    Args:
+        vtof_msgs: Array of shape (num_edge_states,).
+            This holds all the flattened (binary) variables to factor messages.
+        parents_states: Array of shape (num_parents, 2)
+            parents_states[ii, 0] contains the global factor index
+            parents_states[ii, 1] contains the message index of the parent variable's state 0.
+            The parent variable's state 1 is parents_states[ii, 2] + 1
+        children_states: Array of shape (num_factors, 2)
+            children_states[ii, 0] contains the global factor index
+            children_states[ii, 1] contains the message index of the child variable's state 0
+            The child variable's state 1 is children_states[ii, 1] + 1
+
+    Returns:
+        Array of shape (num_edge_states,). This holds all the flattened factor to variable messages.
+    """
+    num_top_vars = parents_states.shape[0]
+    num_factors = children_states.shape[0]
+
+    factor_indices = parents_states[..., 0]
+    top_tof_msgs = (
+        vtof_msgs[parents_states[..., 1] + 1] - vtof_msgs[parents_states[..., 1]]
+    )
+    bottom_tof_msgs = (
+        vtof_msgs[children_states[..., 1] + 1] - vtof_msgs[children_states[..., 1]]
+    )
+
+    def _get_argmaxes(top_tof_msgs, factor_indices):
+        maxes = (
+            jnp.full(shape=num_factors, fill_value=-jnp.inf)
+            .at[factor_indices]
+            .max(top_tof_msgs)
+        )
+        only_maxes_pos = jnp.arange(num_top_vars) - num_top_vars * (
+            top_tof_msgs != maxes[factor_indices]
+        )
+        argmaxes = (
+            jnp.full(shape=num_factors, fill_value=-jnp.inf)
+            .at[factor_indices]
+            .max(only_maxes_pos)
+            .astype(jnp.int32)
+        )
+        return argmaxes
+
+    # Get top incoming argmaxes for each factor
+    first_argmaxes = _get_argmaxes(top_tof_msgs, factor_indices)
+    second_argmaxes = _get_argmaxes(
+        top_tof_msgs.at[first_argmaxes].set(-jnp.inf), factor_indices
+    )
+
+    top_tof_msgs_pos = jnp.maximum(0.0, top_tof_msgs)
+    sum_top_tof_msgs_pos_inc = (
+        jnp.full(shape=num_factors, fill_value=0.0)
+        .at[factor_indices]
+        .add(top_tof_msgs_pos)
+    )
+
+    # Outgoing messages to top variables
+    top_msgs = jnp.minimum(
```

Comment:
I'm getting confused about this update. Shouldn't the update be (for non-firstargmaxes):
```
top_msgs_to1 = bottom_tof_msgs[factor_indices] + sum_top_tof_msgs_pos_inc[factor_indices] - top_tof_msgs_pos # The given parent is 1, child has to be 1, other parents can take whatever value they want to take
top_msgs_to0 = jnp.maximum(
	0.0, # The given parent is 0, child is 0, all other parents are 0
	bottom_tof_msgs[factor_indices] + sum_top_tof_msgs_pos_inc[factor_indices] - top_tof_msgs_pos + jnp.minimum(0.0, top_tof_msgs[first_argmaxes][factor_indices]) # The given parent is 0, child is 1, at least one other parent needs to be 1
)
top_msgs = top_msgs_to1 - top_msgs_to0
```
Are these equivalent?

Reply:
Actually they are equivalent :+1: Might be better to include a bit of comments here (e.g. just copying the above more naive implementation as comments)